autorest/azure/auth/auth.go

```diff
@@ -60,126 +75,181 @@ func NewAuthorizerFromEnvironment() (autorest.Authorizer, error) {
 // 3. Username password
 // 4. MSI
 func NewAuthorizerFromEnvironmentWithResource(resource string) (autorest.Authorizer, error) {
-	settings, err := getAuthenticationSettings()
+	settings, err := GetSettingsFromEnvironment()
 	if err != nil {
 		return nil, err
 	}
-	settings.resource = resource
-	return settings.getAuthorizer()
-}
-
-type settings struct {
-	tenantID            string
-	clientID            string
-	clientSecret        string
-	certificatePath     string
-	certificatePassword string
-	username            string
-	password            string
-	envName             string
-	resource            string
-	environment         azure.Environment
-}
-
-func getAuthenticationSettings() (s settings, err error) {
-	s = settings{
-		tenantID:            os.Getenv("AZURE_TENANT_ID"),
-		clientID:            os.Getenv("AZURE_CLIENT_ID"),
-		clientSecret:        os.Getenv("AZURE_CLIENT_SECRET"),
-		certificatePath:     os.Getenv("AZURE_CERTIFICATE_PATH"),
-		certificatePassword: os.Getenv("AZURE_CERTIFICATE_PASSWORD"),
-		username:            os.Getenv("AZURE_USERNAME"),
-		password:            os.Getenv("AZURE_PASSWORD"),
-		envName:             os.Getenv("AZURE_ENVIRONMENT"),
-		resource:            os.Getenv("AZURE_AD_RESOURCE"),
-	}
-
-	if s.envName == "" {
-		s.environment = azure.PublicCloud
+	settings.Values[Resource] = resource
+	return settings.GetAuthorizer()
+}
+
+// EnvironmentSettings contains the available authentication settings.
+type EnvironmentSettings struct {
+	Values      map[string]string
+	Environment azure.Environment
+}
+
+// GetSettingsFromEnvironment returns the available authentication settings from the environment.
+func GetSettingsFromEnvironment() (s EnvironmentSettings, err error) {
+	s = EnvironmentSettings{
+		Values: map[string]string{},
+	}
+	s.setValue(SubscriptionID)
+	s.setValue(TenantID)
+	s.setValue(ClientID)
+	s.setValue(ClientSecret)
+	s.setValue(CertificatePath)
+	s.setValue(CertificatePassword)
+	s.setValue(Username)
+	s.setValue(Password)
+	s.setValue(EnvironmentName)
+	s.setValue(Resource)
+	if v := s.Values[EnvironmentName]; v == "" {
+		s.Environment = azure.PublicCloud
 	} else {
-		s.environment, err = azure.EnvironmentFromName(s.envName)
+		s.Environment, err = azure.EnvironmentFromName(v)
+	}
+	if s.Values[Resource] == "" {
+		s.Values[Resource] = s.Environment.ResourceManagerEndpoint
 	}
 	return
 }
 
-func (settings settings) getAuthorizer() (autorest.Authorizer, error) {
+// GetSubscriptionID returns the available subscription ID or an empty string.
+func (settings EnvironmentSettings) GetSubscriptionID() string {
+	return settings.Values[SubscriptionID]
+}
+
+// adds the specified environment variable value to the Values map if it exists
+func (settings EnvironmentSettings) setValue(key string) {
```

Comment:
not sure if its relevant in this case, but if we need to distinguish between unset value and an empty value that is set we can use `LookupEnv` (https://golang.org/pkg/os/#LookupEnv)